nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,139 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(%{shape: {_m, k} = shape_q} = q, %{shape: {k, n}} = r, opts) do
+    mode = opts[:mode]
+
+    keep_reflectors = mode == :raw
+    calculate_q = mode in [:reduced, :complete, :raw]
+
+    q =
+      if calculate_q do
+        identity(q, shape: shape_q)
+      end
+
+    {q, r, reflectors} =
+      for i <- 0..(n - 1), reduce: {q, r, []} do
+        {q, r, reflectors} ->
+          %{shape: {reflector_n, reflector_m}} =
+            reflector = householder_reflector(r[[i..(k - 1), i]])
+
+          shape_h = {k, k}
+          h = identity(as_type(reflector, r), shape: shape_h)
+
+          h_bin =
+            for col <- 0..(reflector_m - 1), row <- 0..(reflector_n - 1), reduce: to_binary(h) do
+              acc ->
+                set_value_at_index(
+                  acc,
+                  h.type,
+                  shape_h,
+                  [row + i, col + i],
+                  reflector[[row, col]]
+                )
+            end
+
+          h = from_binary(h, h_bin)
+
+          reflectors =
+            if keep_reflectors do
+              [h | reflectors]
+            else
+              reflectors
+            end
+
+          q =
+            if calculate_q do
+              Nx.dot(q, h)
+            end
+
+          r = Nx.dot(h, r)
+
+          {q, r, reflectors}
+      end
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
+
+      :r ->
+        r
+
+      :raw ->
+        {q, r, reflectors}
+    end
+  end
+
+  defp householder_reflector(%{shape: {n}} = a) do
+    a_0 = a[0]
+
+    v =
+      %{type: {_, num_bits} = type_v} =
+      if Nx.sign(a_0) == -1 do
+        Nx.divide(a, Nx.subtract(a_0, Nx.norm(a)))
+      else
+        Nx.divide(a, Nx.add(a_0, Nx.norm(a)))
+      end
+
+    # set first position of v to 1
+    one = scalar_to_binary(1, type_v)
+    v_bin = to_binary(v)
+    <<_::bitstring-size(num_bits), tail::bitstring>> = v_bin
+    v = from_binary(v, one <> tail)
+
+    eye = identity(a, shape: {n, n})
+
+    projection =
+      2
+      |> Nx.divide(Nx.dot(v, v))
+      |> Nx.multiply(Nx.outer(v, v))
+
+    subtract(
+      projection,
+      eye,
+      projection
+    )
+  end
+
+  defp identity(%{type: {_, num_bits} = type} = tensor, shape: {m, n} = shape) do
+    data_size = m * num_bits * n
+    # allocate empty data
+    data = <<0::size(data_size)>>
```

Comment:
Instead of building a full-sized binary of 0s and then changing the indices that are 1s it is much more preferred to create a binary by accumulating.

Building binaries by appending to them is fast as it only requires reallocation when the underlying binary object is at full capacity.

Changing the contents of an exist binary results a totally new binary each time so it's very very slow.

Here are some erlang docs about [constructing binaries](https://erlang.org/doc/efficiency_guide/binaryhandling.html#constructing-binaries)

Reply:
Is there a preference over appending or prepending regarding preformance? I know that for lists prepending is faster.

Reply:
For binaries, appending to an accumulator that has been appended to previously will be faster than just about any other kind of change that can happen to a binary - it results in fewer allocations than other changes.

Reply:
Thanks! That's nice to know